SPM number10

3a4oT opened the pull request:
## Changes in this pull request

 A better version of #1465 =)

- SPM support with script-based generations.

- added macOS Catalyst support



 ### Generate SPM layout

1. From **project's root** run:

   `bash scripts/generate_spm_sources_layout.sh`

  2. Commit Changes

 Repeat those steps each time you delete/add the project's files. **Make sure** to have this CI step which will check that `generate_spm_sources_layout.sh` is not broken.

Issue fixed: #1368 #1406

### Checklist

- [ ] All tests pass. Demo project builds and runs.
- [ ] I added tests, an experiment, or detailed why my change isn't tested.
- [ ] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.
- [ ] I have reviewed the [contributing guide](https://github.com/Instagram/IGListKit/blob/master/.github/CONTRIBUTING.md)

3a4oT commented:
cc @lorixx

3a4oT commented:
I see that job `Facebook Internal - Builds & Tests` failed but it's not very informative. Would be nice to get some feedback :)

lorixx commented:
I checked and we have this failure:
```
Summary: 
In file included from fbobjc/VendorLib/IGListKit/src/Source/IGListKit/IGListAdapter.h:18:
fbobjc/VendorLib/IGListKit/src/Source/IGListKit/IGListAdapterDataSource.h:10:9: fatal error: 'IGListDiffable.h' file not found
#import "IGListDiffable.h"
        ^~~~~~~~~~~~~~~~~~
In file included from fbobjc/VendorLib/IGListKit/src/Source/IGListKit/IGListExperimentalAdapterUpdater.m:17:
fbobjc/VendorLib/IGListKit/src/Source/IGListDiffKit/Internal/IGListIndexSetResultInternal.h:10:9: fatal error: 'IGListIndexSetResult.h' file not found
#import "IGListIndexSetResult.h"
        ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from fbobjc/VendorLib/IGListKit/src/Source/IGListKit/Internal/IGListBindingSectionController+DebugDescription.h:8:
fbobjc/VendorLib/IGListKit/src/Source/IGListKit/IGListBindingSectionController.h:10:9: fatal error: 'IGListMacros.h' file not found
#import "IGListMacros.h"
        ^~~~ ...(14450 bytes)... 0:03:09,511][vcsutils.mercurial][DEBUG] run hg: hg id -i
=================================

```

RamblinWreck77 commented:
Hey guys! Any updates? I'd love to get this out on the main repo as soon as possible. For reference, this branch has been working great for me in the latest xcode:

https://github.com/3a4oT/IGListKit/tree/spmBrain

We even took a snapshot and pushed it to prod with zero issues/crashes/reported problems.

3a4oT commented:
Hey! this is on my to-do list (upcoming, maybe next weekend :)), but I'm not sure how to deal with internal tests that are failing for Instagram.

3a4oT commented:
OK, so I rebased with the current master! Seems like at least locally I still can build it as a swift package which is good. ~3 months later I see that the current master:
-  has broken Cocoapods and Carthage support
-  IGListDiffKit is no longer a macOS framework since it has "UIKIt" import
-  Some unit tests are failing locally
(I can say it because I have separate PR for GitHub-action based CI which was all green while then)
Current status of this PR:

From the error description `fbobjc/VendorLib/IGListKit/src/Source/`  I can assume that Facebook is using some internal build tool and if that's correct it makes things a little harder to debug. I guess errors like `fbobjc/VendorLib/IGListKit/src/Source/IGListKit/IGListAdapterDataSource.h:10:9: fatal error: 'IGListDiffable.h' file not found
#import "IGListDiffable.h"` can be fixed by conditional imports but I need a way to trigger CI. Some side question, does the Instagram team interest to accept the third-party changes and work with contributors? I ask because :

- you do not have a clear release schedule (at least for me)
- you do not have a community available CI
- this GitHub repository seems to be more like a mirror?

Any answer is good for me, I just need to know whether I should continue to invest time and try to make upstream or just create a hard fork for my own needs?=)

Thanks for the awesome framework :)

3a4oT commented:
https://github.com/Instagram/IGListKit/pull/1478 introduce CI jobs, would be great to land it first. @lorixx

BrentMifsud commented:
any updates on SPM support? Its still broken for me.

lorixx commented:
Thanks for raising the issue @3a4oT:

- We def dont want to break the CI, however the Travis CI was deprecated mid last year and we hadn't prioritized fixing it. But I think we should use the Github Action! Thanks for your initiative to spin up the effort in another PR!
- We would love to resolve these two: Add Github Action CI and support SPM as the community keeps growing and the SPM becomes mainstream.
- I dont think the Github repo is a minor here, would def love to see this getting better and better!

Please keep the feedback and suggestions coming!

lorixx commented:
Awesome @3a4oT ! Looks like all the Unit Tests + Build are done successfully! Do you mind rebase withe the master branch instead? It already includes some of the changes: e.g. Github Action.

Can you create a separate PR for fixing the CI? Thanks a lot!

lorixx commented:
This is the most heroic PR the community got recently! Thanks Petro! @3a4oT

3a4oT commented:
@lorixx I rebased on master. 

- moved most of the CI jobs to the latest macOS and Xcode 12.5.1
- UITests failing on CI under macOS11, so I left it under macOS 10.15 and the latest available Xcode - 12.4

lorixx commented:
> @lorixx I rebased on master(please double-check since there was a mess with commits and I completely forgot what I did half a year ago :)).
> 
> * moved most of the CI jobs to the latest macOS and Xcode 12.5.1
> * UITests failing on CI under macOS11, so I left it under macOS 10.15 and the latest available Xcode - 12.4
> * Added jobs that verify SPM targets can be built by an Xcode. RE. your "test plan". Run "swift build" directly is wrong for this package since SPM builds for the host environment and can't target anything other than macOS. The proper way to do so is to remove the project's `*.xcproject/workspace` files. And invoke `xcodebuild ` tool which under the hood will use Package.swift for a build.
> * However, the above method does not guarantee that third-party apps can consume IGListKit without issues, we may need to add an Example project which integrates with the local version of IGListKit and build it as a CI job.
> 
> Does it all make sense for you?

Thanks @3a4oT ! It all makes sense to me, I will trigger the internal build and ensure the latest didn't break internally as well then we can land the change. Finger cross! The only thing I would ask your help in the gigantic PR change, is to add comment in the change that you believe worth mentioning: e.g. Pod file change, special handling by changing the bitcode enabled or any workaround that applies to the swift package manager layout.

I might ask question inline the PR too.

Again, thank you so much for your hard work here!

lorixx commented on Source/IGListKit/IGListAdapterUpdaterCompatible.h:
```diff
@@ -0,0 +1,91 @@
+/*
```
Please remove this file, it's not in master anymore

3a4oT replied:
done

lorixx commented on scripts/generate_spm_sources_layout.sh:
```diff
@@ -0,0 +1,224 @@
+
+#  2. Commit Changes
+
+#  Repeate those steps each time you delete/add project's files. **Make sure** to have this CI step which will check that `generate_spm_sources_layout.sh` is not broken.
```
Looks like we dont have this step in CI yet, maybe add it to the Github action file or the ./setup.sh too? What do you think?

lorixx commented on Source/IGListDiffKit/IGListBatchUpdateData.mm:
```diff
@@ -7,10 +7,14 @@
 
 #import "IGListBatchUpdateData.h"
 
+#if __has_include(<UIKit/UIKit.h>)
+#import <UIKit/UIKit.h>
+#endif
+
```
Remove this please

lorixx commented on Source/IGListDiffKit/IGListDiff.mm:
```diff
@@ -7,11 +7,15 @@
 
 #import "IGListDiff.h"
 
+#if __has_include(<UIKit/UIKit.h>)
+#import <UIKit/UIKit.h>
+#endif
+
```
Remove this

lorixx commented on Podfile:
```diff
@@ -13,3 +13,15 @@ target 'IGListKit-tvOSTests' do
     pod 'OCMock', '~> 3.8.1'
 end
+
+post_install do |installer|
+    puts "Running OCMock bitcode workaround. https://github.com/erikdoe/ocmock/issues/475"
+    installer.pods_project.targets.each do |target|
+        target.build_configurations.each do |config|
+            # Disable bitcode for OCMock. See: https://github.com/erikdoe/ocmock/issues/475
+            if target.name.start_with?('OCMock')
+                config.build_settings['ENABLE_BITCODE'] = 'NO'
+              end
+        end
+    end
+end
```
Is this the workaruond that fixed the iOS test failure from CI? See my previous attempt that kept failing to open a Pod file: /Users/runner/work/IGListKit/IGListKit/Pods/Target Support Files/OCMock-iOS/OCMock-iOS.debug.xcconfig: unable to open file (in target "OCMock-iOS" in project "Pods") (in target 'OCMock-iOS' from project 'Pods')

https://github.com/Instagram/IGListKit/runs/3345585327?check_suite_focus=true

3a4oT replied:
That was my first idea, so I applied this workaround, but it is not needed and I will remove it, thanks. The issue was that you updated OCMock but didn't commit all "new" files.

lorixx reviewed (changes requested):
Overall it looks great! The next step is to address my inline comment and document anything tricky workaround you did in the development. At the same time, I will work on the internal build to ensure the build works. Thanks Petro!

3a4oT commented:
@lorixx I addressed your feedback, please let me know whether I need to do something else

lorixx commented:
Congrats @3a4oT! Finally merged the change, SPM support should be now available! Btw, do you have idea why https://github.com/Instagram/IGListKit/runs/3479832142?check_suite_focus=true still has unit tests build issue? Thanks!
